airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/concurrent/adapters.py

```diff
@@ -0,0 +1,323 @@
+#
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+#
+
+import copy
+import json
+import logging
+from functools import cache
+from typing import Any, Iterable, List, Mapping, MutableMapping, Optional, Tuple, Union
+
+from airbyte_cdk.models import AirbyteLogMessage, AirbyteMessage, AirbyteStream, Level, SyncMode, Type
+from airbyte_cdk.sources import AbstractSource, Source
+from airbyte_cdk.sources.connector_state_manager import ConnectorStateManager
+from airbyte_cdk.sources.file_based.availability_strategy import AbstractFileBasedAvailabilityStrategy
+from airbyte_cdk.sources.file_based.config.file_based_stream_config import FileBasedStreamConfig
+from airbyte_cdk.sources.file_based.file_types.file_type_parser import FileTypeParser
+from airbyte_cdk.sources.file_based.remote_file import RemoteFile
+from airbyte_cdk.sources.file_based.schema_validation_policies import AbstractSchemaValidationPolicy
+from airbyte_cdk.sources.file_based.stream import AbstractFileBasedStream
+from airbyte_cdk.sources.file_based.stream.concurrent.cursor import FileBasedNoopCursor
+from airbyte_cdk.sources.message import MessageRepository
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.concurrent.default_stream import FileBasedDefaultStream
+from airbyte_cdk.sources.streams.concurrent.exceptions import ExceptionWithDisplayMessage
+from airbyte_cdk.sources.streams.concurrent.helpers import get_cursor_field_from_stream, get_primary_key_from_stream
+from airbyte_cdk.sources.streams.concurrent.partitions.partition import Partition
+from airbyte_cdk.sources.streams.concurrent.partitions.partition_generator import PartitionGenerator
+from airbyte_cdk.sources.streams.concurrent.partitions.record import Record
+from airbyte_cdk.sources.streams.core import StreamData
+from airbyte_cdk.sources.utils.schema_helpers import InternalConfig
+from airbyte_cdk.sources.utils.slice_logger import SliceLogger
+from deprecated.classic import deprecated
+
+"""
+This module contains adapters to help enabling concurrency on File-based Stream objects without needing to migrate to AbstractStream
+"""
+
+
+@deprecated("This class is experimental. Use at your own risk.")
+class FileBasedStreamFacade(Stream):
```

Comment:
I hear you on the duplicated code. I'm unable to use `StreamFacade` here because there are file-specific methods and types that are required by the callers. 

I've played around a lot with this to try to clean it up - e.g. creating an interface that `StreamFacade` and `FileBasedStreamFacade` can use, and avoiding the use of `StreamFacade`-like pattern altogether, and avoiding use of `DefaultStream`, but did not come up with something that I felt worked with the existing framework.

What I'd like to do is keep this as-is, and consider a separate refactor that would help make the concurrency interface a little more flexible, but I'd prefer not to block this on it. In the mean time I'm not worried about the fact that we have a parallel file-based flow that has some duplicated code since it does still fit into the concurrency patterns that we've put in place rather than reinventing it from scratch.

Reply:
@clnoll can you create an issue specifying exactly what problems the refactor should resolve so we don't lose track of it?

The bit I'm particularly surprised about is that the file-based classes need to know about the stream facade. Maybe `FileBasedStreamFacade` should implement a `AbstractFileBasedStream`? This way, the mess is kept to the facade and the file-based framework doesn't know anything about it

Reply:
Update: I was at least able to git rid of the need for `DefaultFileBasedStream`.

Reply:
@girarda I wrote that last comment before I saw yours.

Good call re having `FileBasedStreamFacade` implement `AbstractFileBasedStream`, made that update and it cleaned up some of the issues with types.